### PR TITLE
Improve Fuzz CI

### PR DESCRIPTION
**Pull request by RCasatta**
My original plan was to run fuzz testing in CI separately with both x86_64 and i686 arch.

Unfortunately, I got issues with the i686 fuzzing https://github.com/RCasatta/rust-bitcoin/actions/runs/1265153484 and apparently specifying different targets is not supported https://github.com/rust-fuzz/honggfuzz-rs/blob/f45aef1b59736fcfc215eca09aca42e49bbf36db/src/bin/cargo-hfuzz.rs#L23

Nevertheless, I think there are some advantages in doing fuzzing in CI separately (faster, better report), however, there are some disadvantages (what launched from the CI is not exactly the same script launched locally). For this reason, I am not sure if 3e310d3c261335e2cda919a3c261b6591ca0e32e is desirable and if yes, if we want to remove fuzzing in the nightly test.

In case we keep the fuzzing in the nightly test I think 9049eef7004e1f52f19352ce0cc362d152fe3a11 is desirable because it saves about 30 seconds for other tests that don't need those deps.

### CODE REVIEW

File: .github/workflows/fuzz.yml [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+jobs:
+
+  fuzz:
+    runs-on: ubuntu-20.04
+    env:
+      HFUZZ_BUILD_ARGS: "--features honggfuzz_fuzz"
+    strategy:
+      fail-fast: false
+      matrix:
+        fuzz_target: [deser_net_msg, deserialize_address, deserialize_amount, deserialize_block, deserialize_psbt, deserialize_script, deserialize_transaction, outpoint_string, uint128_fuzz]
```
Comment (TheBlueMatt) [inline .github/workflows/fuzz.yml:14]: If we can't generate this from the file list, can CI somewhere fail if this misses a fuzz target?

File: .github/workflows/rust.yml [hunk 2]
```diff
@@ -29,8 +29,11 @@ jobs:
               AS_DEPENDENCY: true
               PIN_VERSIONS: true
     steps:
-      - name: Install test dependencies
+
+      - name: Install fuzz dependencies
```
Comment (TheBlueMatt) [inline .github/workflows/rust.yml:33]: Can't this be dropped entirely now?

File: .github/workflows/fuzz.yml (hunk 1 above)
Comment (RCasatta) [reply .github/workflows/fuzz.yml:14]: yes, check done in 1aefc1ccf3555d707de39ee31238046aab8c4b47

File: .github/workflows/rust.yml (hunk 2 above)
Comment (RCasatta) [reply .github/workflows/rust.yml]: yes, done in 9b6b50a9870608bff9e7ecf8042e83cdb09cf640

File: .github/workflows/fuzz.yml [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    needs: fuzz
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+      - uses: actions/download-artifact@v2
+      - name: Display structure of downloaded files
+        run: ls -R
+      - run: find executed_* -type f -exec cat {} + | sort > executed
+      - run: ls fuzz/fuzz_targets | sort > expected
+      - run: diff expected executed
```
Comment (TheBlueMatt) [inline .github/workflows/fuzz.yml:53]: I think you want some kind of if statement that if there's a diff we fail the test.

**Review by TheBlueMatt — COMMENTED** [review]
LGTM, mod one comment.

File: .github/workflows/fuzz.yml (hunk 3 above)
Comment (RCasatta) [reply .github/workflows/fuzz.yml:53]: The if isn't necessary, `diff` returns not 0 if files are different. See my test where I explicitly removed one fuzz target https://github.com/RCasatta/rust-bitcoin/runs/3698414243?check_suite_focus=true

**Comment by dr-orlovsky** [comment]
Why do we have such a big diff with all temporary files (which should never get into git) removed?

**Comment by RCasatta** [comment]
> Why do we have such a big diff with all temporary files (which should never get into git) removed?

> which should never get into git

Those files are fuzzing test cases stimulating different program paths. Fuzzers could recreate those by running and searching for various inputs, but it could take a long time. Providing those to the CI ensures many program paths are tested even if the fuzzing runs for just 30 seconds, that's why they are committed into git.

> Why removed?

fuzz targets `deserialize_decimal` and `deserialize_udecimal` are not there anymore (maybe renamed?) so that files are not used anymore.

**Comment by RCasatta** [comment]
Using rust stable toolchain since honggfuzz is supporting it -> https://github.com/rust-bitcoin/rust-bitcoin/issues/667#issuecomment-928096644

**Review by apoelstra — APPROVED** [review]
ACK 2bbf63c7e0ba4901ecec84bb2118fe166f8154fe

**Review by dr-orlovsky — APPROVED** [review]
utACK 2bbf63c7e0ba4901ecec84bb2118fe166f8154fe

Awaiting for @TheBlueMatt confirmation that he is satisfied on the answer to his comment and after that will merge (or he can merge) to get CI fixed and unlock further merging of other PRs

**Review by TheBlueMatt — APPROVED** [review]
Oops, sorry, yes, my comment was resolved.

**Pull request merged**